Kick the tires on a documentation cheatsheet (#867)
lalrpop/lalrpop: doc/src/cheatsheet.md

# Cheatsheet

> Users of Lalrpop has compiled the following cheatsheet table as a quick way to

Reviewer: ```suggestion
Users of Lalrpop have compiled the following cheatsheet table as a quick way to
```

# Cheatsheet

Users of Lalrpop has compiled the following cheatsheet table as a quick way to
look up useful Lalrpop-isms. If you are looking for a specific piece of
functionality, use this table to jump to the right section.

| name | snipped | description | tutorial |
|---|---|---|---|

> | position | `<@L> T <@R>` | uhm, I think it gives the index of the parser in the input | - |

Reviewer: Not that the original example doesn't work, but let's

```suggestion
| position | `<left: @L> T <right: @R>` | captures the byte offset of the first byte and the offset of the last byte plus one (as `left` and `right` respectively) | - |
```

> | error_recovery | `! => { ... }` | recover from parser errors | [Error recovery](tutorial/008_error_recovery.md) |

Reviewer: ```suggestion
| error_recovery | `! => { ... }` | recovers from parser errors | [Error recovery](tutorial/008_error_recovery.md) |
```

| grammar_parameter | `grammar(scale: isize);` | input parameters usable in the generated parser | [Passing state parameter](tutorial/009_state_parameter.md) |

> | ?? | `Num => func(<i32>)` | maybe automatic number parsing?? | - |

Reviewer: What does this one do, exactly? Do you have a complete example? Because I'm not sure to know the feature you're referring to

Author: Yeah, I'm not sure what the original poster meant by this the more I look into it. The closest I was able to find was something like:
```
extern {
    enum Tok<'input> {
        "(" => Tok::LParen,
        ")" => Tok::RParen,
        "-" => Tok::Minus,
        "+" => Tok::Plus,
        "*" => Tok::Times,
        "/" => Tok::Div,
        Num => Tok::Num(<i32>),
        Fraction => Tok::Fraction(<i32>, <i32>), // Regression test for #179
    }
}
```

Reviewer: Oh, I see. I believe it's the usual way to map terminals to lexer tokens when using a custom lexer. You have to specify the type of data that the token carries (and that you want to extract in the parser): this just means that `Fraction` is defined as `enum Tok { ..., Fraction(i32, i32), ... }` in the lexer, which will correspond to the non-terminal `Fraction` that will have type `(i32, i32)` in the parser.

> | custom_error | `"e" =>? ParseError::User { error: "an error" }` | make an action failable | [Fallible actions](tutorial/007_fallible_actions.md) |

Reviewer: ```suggestion
| custom_error | `"e" =>? Err(ParseError::User { error: "an error" })` | makes an action fallible | [Fallible actions](tutorial/007_fallible_actions.md) |
```

> | custom_macros | `Comma<T> =  { ... }` | allows a non-terminal to be generic about some parameters | [Macros](tutorial/006_macros.md) |

Reviewer: ```suggestion
| custom_macros | `Comma<T> =  { ... }` | makes a non-terminal generic in other non-terminals | [Macros](tutorial/006_macros.md) |
```

> | quantifier_macros | `<Num?> <Num*> <Num+>` |  a non-terminal which can appear 0..1, 0+, 1+ times | [Macros](tutorial/006_macros.md) |

Reviewer: I don't think those are really called macros somewhere in LALRPOP's documentation, are they?

```suggestion
| quantifier | `<Num?> <Num*> <Num+>` |  a non-terminal which can appear 0..1, 0+, 1+ times | [Macros](tutorial/006_macros.md) |
```

Author: The documentation linked says: 
```
In fact, LALRPOP comes with four macros builtin: `*`, `?`, `+`, and `(...)`.
```

I am happy with just calling them quantifiers though to be short and clear.

Reviewer: Oh, I missed that. Then it makes sense to stay consistent here :+1:

> | tuple_macro | `<a:(<Num> ",")*>` | when used with a qunatifier allows crating lists | [Macros](tutorial/006_macros.md) |

Reviewer: ```suggestion
| tuple_macro | `<a:(<Num> ",")*>` | applies a quantifier to a group of matches | [Macros](tutorial/006_macros.md) |
```

| extern | `extern { }` | allows to override some parts of the generated parser | [Writing a custom lexer](lexer_tutorial/003_writing_custom_lexer.md) |
| extern_error | `type Error = MyError;` | sets the error to use in the `ParseError::User` variant | [Writing a custom lexer](lexer_tutorial/003_writing_custom_lexer.md) |

> | extern_location | `type Location = MyLoc;` | set the type to for locations instead of `usize` | [Writing a custom lexer](lexer_tutorial/003_writing_custom_lexer.md) |

Reviewer: ```suggestion
| extern_location | `type Location = MyLoc;` | sets the type to for locations instead of `usize` | [Writing a custom lexer](lexer_tutorial/003_writing_custom_lexer.md) |
```

> | extern_tok | `enum Tok { }` | set the type to use for tokens?? | [Using tokens with references](lexer_tutorial/004_token_references.md) |

Reviewer: Nitpick, but just to make it clear that `Tok` isn't a magic symbol and we can actually use whatever custom token type we've defined elsewhere in our custom lexer.

```suggestion
| extern_tok | `enum MyToken { }` | declare the type of lexer tokens to be consumed by the generated parser  | [Using tokens with references](lexer_tutorial/004_token_references.md) |
```

> | auto_parameters | `<>` | in an action it turns into a list of parameters | [Type inference](tutorial/003_type_inference.md) |

Reviewer: ```suggestion
| auto_parameters | `<>` | refers to all the parameters of the non-terminal as a tuple | [Type inference](tutorial/003_type_inference.md) |
```

> |conditional actions | `<T> if T == "a" => (),` | - | - |

Reviewer: Although conditional macro are currently undocumented, I believe the idea is to have a macro parameter (ab)used as a string constant, but usually you don't use this parameter directly in the alternative (see https://gist.github.com/nikomatsakis/5fa3bd8291841b853144)

```suggestion
|conditional actions | `Expr<I> = ... , <T> if I == "a" => (),` | Conditional definition of a macro's alternative | - |
```
